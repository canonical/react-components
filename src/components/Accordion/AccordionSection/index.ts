export { default } from "./AccordionSection";
export type {
  Props as AccordionSectionProps,
  AccordionHeadings,
} from "./AccordionSection";
