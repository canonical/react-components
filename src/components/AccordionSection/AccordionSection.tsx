import React from "react";
import type { ReactNode } from "react";

import type { Headings } from "types";
import { useId } from "hooks/useId";

export type AccordionHeadings = Exclude<Headings, "h1">;

export type Props = {
  /**
   * The content of the section.
   */
  content?: ReactNode;
  /**
   * An optional value to set the expanded section. The value must match a
   * section key.
   */
  expanded?: string;
  headingLevel?: number;
  /**
   * An optional click event when the title is clicked.
   */
  onTitleClick?: (expanded: boolean, key: string) => void;
  /**
   * An optional key to be used to track which section is selected.
   */
  sectionKey?: string;
  setExpanded?: (key: string | null, title: string | null) => void;
  /**
   * The title of the section.
   */
  title?: string;
  /**
   * Optional string describing heading element that should be used for the section titles.
   */
  titleElement?: AccordionHeadings;
};

const AccordionSection = ({
  content,
  expanded,
  onTitleClick,
  sectionKey,
  setExpanded,
  title,
  titleElement,
  headingLevel = 3,
}: Props): JSX.Element => {
  const sectionId = useId();
  const key = sectionKey || sectionId;
  const isExpanded = expanded === key;
  const Title = titleElement || "div";

  return (
    <li className="p-accordion__group">
      <Title
        role={titleElement ? null : "heading"}
        aria-level={titleElement ? null : headingLevel}
        className="p-accordion__heading"
      >
        <button
          aria-controls={`#${sectionId}`}
          aria-expanded={isExpanded ? "true" : "false"}
          className="p-accordion__tab"
          onClick={() => {
            if (isExpanded) {
              setExpanded(null, null);
            } else {
              setExpanded(key, title);
            }
            onTitleClick && onTitleClick(!isExpanded, key);
          }}
          role="tab"
          type="button"
        >
          {title}
        </button>
      </Title>
      <section
        aria-hidden={isExpanded ? "false" : "true"}
        aria-labelledby={sectionId}
        className="p-accordion__panel"
        id={sectionId}
        role="tabpanel"
      >
        {content}
      </section>
    </li>
  );
};

export default AccordionSection;
