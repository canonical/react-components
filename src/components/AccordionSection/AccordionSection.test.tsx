import { shallow } from "enzyme";
import React from "react";

import AccordionSection from "./AccordionSection";

describe("AccordionSection ", () => {
  it("renders", () => {
    const wrapper = shallow(
      <AccordionSection
        content={<span>Test</span>}
        expanded="abcd-1234"
        setExpanded={jest.fn()}
        title="Test section"
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("renders headings for titles", () => {
    const wrapper = shallow(
      <AccordionSection
        content={<span>Test</span>}
        expanded="abcd-1234"
        setExpanded={jest.fn()}
        title="Test section"
        titleElement="h4"
      />
    );

    expect(wrapper.find("h4.p-accordion__heading")).toHaveLength(1);
    expect(wrapper.find("h4").first().prop("aria-level")).toBeNull();
    expect(wrapper).toMatchSnapshot();
  });

  it("can handle click events on the title", () => {
    const onTitleClick = jest.fn();
    let expanded = null;
    const wrapper = shallow(
      <AccordionSection
        content={<span>Test</span>}
        expanded={expanded}
        onTitleClick={onTitleClick}
        setExpanded={(id) => {
          expanded = id;
        }}
        title="Test section"
      />
    );
    wrapper.find(".p-accordion__tab").at(0).simulate("click");

    expect(onTitleClick.mock.calls[0]).toEqual([true, "mock-nanoid-1"]);
    wrapper.setProps({ expanded });
    // Clicking the title again should close the accordion section.
    wrapper.find(".p-accordion__tab").at(0).simulate("click");

    expect(onTitleClick.mock.calls[1]).toEqual([false, "mock-nanoid-1"]);
  });

  it("can use a key for expanded state", () => {
    const onTitleClick = jest.fn();
    let expanded = null;
    const wrapper = shallow(
      <AccordionSection
        content={<span>Test</span>}
        expanded={expanded}
        sectionKey="first-key"
        onTitleClick={onTitleClick}
        setExpanded={(id) => {
          expanded = id;
        }}
        title="Test section"
      />
    );
    wrapper.find(".p-accordion__tab").at(0).simulate("click");

    expect(onTitleClick.mock.calls[0]).toEqual([true, "first-key"]);
  });

  it("can use a key for expanded state", () => {
    const onTitleClick = jest.fn();
    let expanded = null;
    const wrapper = shallow(
      <AccordionSection
        content={<span>Test</span>}
        expanded={expanded}
        sectionKey="first-key"
        onTitleClick={onTitleClick}
        setExpanded={(id) => {
          expanded = id;
        }}
        title="Test section"
      />
    );
    wrapper.find(".p-accordion__tab").at(0).simulate("click");

    expect(onTitleClick.mock.calls[0]).toEqual([true, "first-key"]);
  });
});
