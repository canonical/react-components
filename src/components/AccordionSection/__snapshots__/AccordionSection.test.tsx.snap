// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionSection  renders 1`] = `
<li
  className="p-accordion__group"
>
  <div
    aria-level={3}
    className="p-accordion__heading"
    role="heading"
  >
    <button
      aria-controls="#mock-nanoid-1"
      aria-expanded="false"
      className="p-accordion__tab"
      onClick={[Function]}
      role="tab"
      type="button"
    >
      Test section
    </button>
  </div>
  <section
    aria-hidden="true"
    aria-labelledby="mock-nanoid-1"
    className="p-accordion__panel"
    id="mock-nanoid-1"
    role="tabpanel"
  >
    <span>
      Test
    </span>
  </section>
</li>
`;

exports[`AccordionSection  renders headings for titles 1`] = `
<li
  className="p-accordion__group"
>
  <h4
    aria-level={null}
    className="p-accordion__heading"
    role={null}
  >
    <button
      aria-controls="#mock-nanoid-1"
      aria-expanded="false"
      className="p-accordion__tab"
      onClick={[Function]}
      role="tab"
      type="button"
    >
      Test section
    </button>
  </h4>
  <section
    aria-hidden="true"
    aria-labelledby="mock-nanoid-1"
    className="p-accordion__panel"
    id="mock-nanoid-1"
    role="tabpanel"
  >
    <span>
      Test
    </span>
  </section>
</li>
`;
